api/modelgeneration/modelgeneration_test.go

```diff
@@ -0,0 +1,136 @@
+// Copyright 2019 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package modelgeneration_test
+
+import (
+	"github.com/golang/mock/gomock"
+	gc "gopkg.in/check.v1"
+	"gopkg.in/juju/names.v2"
+
+	"github.com/juju/juju/api/base/mocks"
+	"github.com/juju/juju/api/modelgeneration"
+	"github.com/juju/juju/apiserver/params"
+	"github.com/juju/juju/core/model"
+)
+
+type modelGenerationSuite struct {
+	tag names.ModelTag
+}
+
+var _ = gc.Suite(&modelGenerationSuite{})
+
+func (s *modelGenerationSuite) SetUpTest(c *gc.C) {
+	s.tag = names.NewModelTag("deadbeef-abcd-4fd2-967d-db9663db7bea")
+}
+
+func (s *modelGenerationSuite) TestAddGeneration(c *gc.C) {
+	ctrl := gomock.NewController(c)
+	defer ctrl.Finish()
+
+	resultSource := params.ErrorResult{}
+	arg := params.Entity{Tag: s.tag.String()}
+
+	caller := mocks.NewMockAPICallCloser(ctrl)
```

Comment:
The most elegant pattern I've found for these repeated blocks of controller set-up and base assertions is to make the mock(s) a member of the suite and use a method like:
```go
func (s *modelGenerationSuite) setUpMocks(c *gc.C) {
    ctrl := gomock.NewController(c)
    
    s.Caller := mocks.NewMockAPICallCloser(ctrl)
    cExp := s.Caller.EXPECT()
    cExp.BestFacadeVersion(gomock.Any()).Return(0).AnyTimes()

    s.FacadeCaller := mocks.NewMockFacadeCaller(ctrl)
    fExp := s.FacadeCaller.EXPECT()
    fExp.RawAPICaller().Return(caller).AnyTimes()
    
    return ctrl
}
```
Then add the top of the tests, use:
```go
    defer s.setupMocks(c).Finish()
```

Reply:
nice. thank you.